src/geometry/LinearPropagator.i.hh

```diff
@@ -0,0 +1,62 @@
+//---------------------------------*-CUDA-*----------------------------------//
+// Copyright 2020 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file LinearPropagation.i.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "geometry/LinearPropagator.hh"
+#include "base/ArrayUtils.hh"
+#include "geometry/Types.hh"
+#include "physics/base/Units.hh"
+
+namespace celeritas
+{
+//___________________________________________________________________________//
+/*!
+ * Do straight propagation to physics process or boundary and reduce next_step
+ *
+ * Scalar geometry length computation. The track is moved along track.dir()
+ * direction by a distance track.next_step()
+ */
+CELER_FORCEINLINE_FUNCTION
+void LinearPropagator::fast_linear_step(real_type step)
+{
+    axpy(step, track_.dir(), &track_.pos());
+    track_.next_step() -= step;
+}
+
+//___________________________________________________________________________//
+/*!
+ * No step provided -> self-aware move to next boundary.
+ *
+ * State gets updated to next volume
+ */
+CELER_FORCEINLINE_FUNCTION
+void LinearPropagator::operator()()
+{
+    this->fast_linear_step(track_.next_step() + track_.tolerance());
+
+    // update state
+    track_.move_next_volume();
```

Comment:
Note: we will very likely have to (re)add support for numerical imprecision.  Sometimes even if you move the particle the distance to the next volume you might still not have reached the boundary (numerically).

See for example GeantV LinearPropagationHandler:

https://gitlab.cern.ch/GeantV/geant/-/blob/master/run/scheduler/src/LinearPropagationHandler.cxx#L28

Reply:
I'd like any numerical imprecision issues to be encapsulated by the geometry wrapper class, and not in the propagator. The `move_next_volume` call inside the geometry track view should guarantee that the track actually makes it to the volume. So actually @mrguilima we should think about whether the track tolerance above gets applied inside the fast linear step, or just when moving across a volume boundary. If we're sub-stepping well inside of a volume, there shouldn't be a need to "bump" the particle.

Reply:
In some circumstances (eg. involving volume overlaps) we found that we even had cases where it did not converge and thus we needed to "kill" the particle in that case.

> >  by the geometry wrapper class, and not in the propagator.

I am not sure what is the responsibility of the propagator then ...

Reply:
I implemented this idea as suggested by @sethrj above, but I did it on the propagator class, see here:
https://github.com/celeritas-project/celeritas/blob/b5ad3a1e0cbf2e774dacc24aeb66e1845e6beb70/src/geometry/LinearPropagator.i.hh#L40

It is not clear to me at this point what @sethrj refers as the 'geometry wrapper class'.  `GeoParams`?

Of course, we may conclude that we still need some special handling for stubborn stuck tracks.  We will see.

Reply:
Sorry for the confusion; I think of "GeoTrackView" as a wrapper class encapsulating the VecGeom API and implementation. (The Shift code templates on a Geometry object so that the same transport algorithms can be agnostic of the geometry implementation, including the need to bump particles, which is handled by the low-level wrapper for the geometry implementations that need it.)